### ISSUE TITLE
[no-unnecessary-condition] False positive with optional chaining on Array methods that have nullish return values

### ISSUE THREAD

**Issue by chimericdream** [typescript-eslint/typescript-eslint#2027]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["warn", {"ignoreRhs": true}]
  }
}
```

```TS
interface SomeThing {
    id: string;
    stuff: string;
}

const arr: SomeThing[] = [/* ... */];

const theRightStuff: string = arr.find(item => item.id === 'right')?.stuff;
```

**Expected Result**
No warning.

**Actual Result**
The linter warns that the optional chain is unnecessary.

**Additional Info**
The return value of Array.prototype.find is `T | undefined`, so I would expect the linter/parser to recognize the output of that function call to be potentially nullish.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.24.0` |
| `@typescript-eslint/parser`        | `2.24.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.2` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
are you using `strictNullChecks`?
If you aren't, then the rule will not work well - our plugin can only consume the types TS has, so if without `strictNullChecks`, `undefined` and `null` are essentially dropped from all types, meaning this rule just sees the type as `SomeThing`, not `SomeThing | undefined`
